cli/azd/pkg/account/subscriptions_cache.go

```diff
@@ -12,88 +14,116 @@ import (
 	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
 )
 
-// The file name of the cache used for storing subscriptions accessible by the currently logged in account.
+// The file name of the cache used for storing subscriptions accessible by local accounts.
 const cSubscriptionsCacheFile = "subscriptions.cache"
 
-// SubscriptionsCache caches the list of subscriptions accessible by the currently logged in account.
+// SubscriptionsCache caches the list of subscriptions accessible by local accounts.
 //
 // The cache is backed by an in-memory copy, then by local file system storage.
+// The cache key should be chosen to be unique to the user, such as the user's object ID.
+//
+// To clear all entries in the cache, call Clear().
 type SubscriptionsCache struct {
-	cachePath string
+	cacheDir string
 
-	inMemoryCopy []Subscription
+	inMemoryCopy map[string][]Subscription
 	inMemoryLock sync.RWMutex
 }
 
-func NewSubscriptionsCache() (*SubscriptionsCache, error) {
+func newSubCache() (*SubscriptionsCache, error) {
 	configDir, err := config.GetUserConfigDir()
 	if err != nil {
 		return nil, fmt.Errorf("loading stored user subscriptions: %w", err)
 	}
 
-	return NewSubscriptionsCacheWithDir(filepath.Join(configDir, cSubscriptionsCacheFile))
-}
-
-func NewSubscriptionsCacheWithDir(cachePath string) (*SubscriptionsCache, error) {
 	return &SubscriptionsCache{
-		cachePath: cachePath,
+		cacheDir:     configDir,
+		inMemoryCopy: map[string][]Subscription{},
 	}, nil
 }
 
-// Load loads the subscriptions from cache. Returns any error reading the cache.
-func (s *SubscriptionsCache) Load() ([]Subscription, error) {
+// Load loads the subscriptions from cache with the key. Returns any error reading the cache.
+func (s *SubscriptionsCache) Load(ctx context.Context, key string) ([]Subscription, error) {
+	// check in-memory cache
 	s.inMemoryLock.RLock()
-	if s.inMemoryCopy != nil {
+	if res, ok := s.inMemoryCopy[key]; ok {
 		defer s.inMemoryLock.RUnlock()
-		return s.inMemoryCopy, nil
+		return res, nil
 	}
 	s.inMemoryLock.RUnlock()
 
 	s.inMemoryLock.Lock()
 	defer s.inMemoryLock.Unlock()
-	cacheFile, err := os.ReadFile(s.cachePath)
+
+	// load cache from disk
+	cacheFile, err := os.ReadFile(filepath.Join(s.cacheDir, cSubscriptionsCacheFile))
 	if err != nil {
 		return nil, err
 	}
 
-	var subscriptions []Subscription
-	err = json.Unmarshal(cacheFile, &subscriptions)
+	var cache map[string][]Subscription
+	err = json.Unmarshal(cacheFile, &cache)
```

Comment:
If the `subscriptions.cache` file is the older version, the unmarshal here will fail, correct?  I see in `Save` we ignore this error and plow ahead (reasonable), do we need to do something here as well or does a larger part of the system handle the fact that `Get` may not return an error even when there are not I/O errors?

Reply:
I see - `GetSubscriptions` is the only caller and on an any error it just ignores the cache and does a full fetch, so this feels good.

Reply:
Made this a non-exported type as  `subscriptionsCache` to make the helper/internal component clearer.